### PR TITLE
Day 6 Julia

### PR DESCRIPTION
**Pull request by Moelf**

### CODE REVIEW

File: src/julia/06_moelf.jl [hunk 1]
```diff
@@ -0,0 +1,25 @@
+boat_distance(hold, total_time) = hold * (total_time - hold)
+
+function ways_to_win(total_time, to_beat)
+    ways = 0
+    for i in 1:total_time
```
Comment (Moelf) [inline src/julia/06_moelf.jl:5]: idk why, but I tried to mark this with `@simd` and it didn't help.

**Comment by Moelf** [comment]
```julia
julia> boat_distance(hold, total) = hold * (total - hold)
boat_distance (generic function with 1 method)

julia> function ways_to_win(total_time, to_beat)
           ways = 0
           for i in 1:total_time
               ways += boat_distance(i, total_time) > to_beat
           end
           return ways
       end
ways_to_win (generic function with 1 method)

julia> using BenchmarkTools

julia> @benchmark ways_to_win(53717880, 275118112151524)
BenchmarkTools.Trial: 590 samples with 1 evaluation.
 Range (min … max):  8.417 ms …   9.958 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     8.449 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   8.486 ms ± 140.872 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▅▅▄▃▃▃▃▃ ▁▁▁▁ ▁▁
  ██████████████████▅▆▁▇▆▁▅▆▄▄▅▆▄▁▆▁▁▁▁▁▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▁▁▄▁▄ ▇
  8.42 ms      Histogram: log(frequency) by time      8.97 ms <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> function ways_to_win(total_time, to_beat)
           ways = 0
           @simd for i in 1:total_time
               ways += boat_distance(i, total_time) > to_beat
           end
           return ways
       end
ways_to_win (generic function with 1 method)

julia> @benchmark ways_to_win(53717880, 275118112151524)
BenchmarkTools.Trial: 560 samples with 1 evaluation.
 Range (min … max):  8.852 ms …  10.415 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     8.894 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   8.935 ms ± 145.179 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▇▆▆▄▃▄▃▂▃ ▁▂  ▁▁
  █████████████████▄▆▅▆▆▄▅█▆▁▅▁▁▁▁▅▁▁▁▄▄▁▁▁▁▁▁▁▁▁▄▄▁▁▁▁▁▄▁▁▁▄ ▇
  8.85 ms      Histogram: log(frequency) by time      9.58 ms <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

**Comment by Moelf** [comment]
ah ok because it's already vectorized:
```asm
julia> @code_native debuginfo=:none ways_to_win(53717880, 275118112151524)
	.text
	.file	"ways_to_win"
	.section	.rodata.cst32,"aM",@progbits,32
	.p2align	5                               # -- Begin function julia_ways_to_win_1069
.LCPI0_0:
	.quad	1                               # 0x1
	.quad	2                               # 0x2
	.quad	3                               # 0x3
	.quad	4                               # 0x4
	.section	.rodata.cst8,"aM",@progbits,8
	.p2align	3
.LCPI0_1:
	.quad	4                               # 0x4
.LCPI0_2:
	.quad	8                               # 0x8
.LCPI0_3:
	.quad	12                              # 0xc
.LCPI0_4:
	.quad	16                              # 0x10
	.text
	.globl	julia_ways_to_win_1069
	.p2align	4, 0x90
	.type	julia_ways_to_win_1069,@function
julia_ways_to_win_1069:                 # @julia_ways_to_win_1069
# %bb.0:                                # %top
	xor	eax, eax
	mov	r9d, 0
	test	rdi, rdi
	cmovg	r9, rdi
	jle	.LBB0_8
# %bb.1:                                # %L17.preheader
	cmp	r9, 16
	jae	.LBB0_3
# %bb.2:
	mov	r10d, 1
	xor	eax, eax
	jmp	.LBB0_6
.LBB0_3:                                # %vector.ph
	push	rbp
	mov	rbp, rsp
	movabs	r8, 9223372036854775792
	and	r8, r9
	lea	r10, [r8 + 1]
	movabs	rax, offset .LCPI0_0
	vmovdqa	ymm0, ymmword ptr [rax]
	vpbroadcastq	ymm1, rdi
	vpbroadcastq	ymm3, rsi
	vpxor	xmm2, xmm2, xmm2
	movabs	rax, offset .LCPI0_1
	vpbroadcastq	ymm4, qword ptr [rax]
	movabs	rax, offset .LCPI0_2
	vpbroadcastq	ymm5, qword ptr [rax]
	movabs	rax, offset .LCPI0_3
	vpbroadcastq	ymm6, qword ptr [rax]
	movabs	rax, offset .LCPI0_4
	vpbroadcastq	ymm7, qword ptr [rax]
	mov	rax, r8
	vpxor	xmm8, xmm8, xmm8
	vpxor	xmm9, xmm9, xmm9
	vpxor	xmm10, xmm10, xmm10
	pop	rbp
	.p2align	4, 0x90
.LBB0_4:                                # %vector.body
                                        # =>This Inner Loop Header: Depth=1
	vpaddq	ymm11, ymm0, ymm4
	vpaddq	ymm12, ymm0, ymm5
	vpaddq	ymm13, ymm0, ymm6
	vpsubq	ymm14, ymm1, ymm0
	vpsubq	ymm15, ymm1, ymm11
	vpsubq	ymm16, ymm1, ymm12
	vpsubq	ymm17, ymm1, ymm13
	vpmullq	ymm14, ymm14, ymm0
	vpmullq	ymm11, ymm15, ymm11
	vpmullq	ymm12, ymm16, ymm12
	vpmullq	ymm13, ymm17, ymm13
	vpcmpgtq	ymm14, ymm14, ymm3
```

**Comment by Moelf** [comment]
that's too smart for me

**Comment by Cryptum169** [comment]
Yeah this is the closed form solution man

**Pull request merged**